src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs

```diff
@@ -0,0 +1,887 @@
+            int dc = component.DcPredictor + diff;
+            component.DcPredictor = dc;
+            blockDataRef = (short)dc;
+
+            // Decode AC Components, See Jpeg Spec
+            int k = 1;
+            do
+            {
+                int zig;
+                int s;
+
+                this.CheckBits();
+                int c = this.PeekBits();
+                int r = Unsafe.Add(ref fastACRef, c);
+
+                if (r != 0)
+                {
+                    // Fast AC path
+                    k += (r >> 4) & 15; // Run
+                    s = r & 15; // Combined Length
+                    this.codeBuffer <<= s;
+                    this.codeBits -= s;
+
+                    // Decode into unzigzag location
+                    zig = this.dctZigZag[k++];
+                    Unsafe.Add(ref blockDataRef, zig) = (short)(r >> 8);
+                }
+                else
+                {
+                    int rs = this.DecodeHuffman(ref acTable);
+
+                    if (rs < 0)
+                    {
+                        throw new ImageFormatException("Bad Huffman code");
+                    }
+
+                    s = rs & 15;
+                    r = rs >> 4;
+
+                    if (s == 0)
+                    {
+                        if (rs != 0xF0)
+                        {
+                            break; // End block
+                        }
+
+                        k += 16;
+                    }
+                    else
+                    {
+                        k += r;
+
+                        // Decode into unzigzag location
+                        zig = this.dctZigZag[k++];
+                        Unsafe.Add(ref blockDataRef, zig) = (short)this.ExtendReceive(s);
+                    }
+                }
+            } while (k < 64);
+        }
+
+        private void DecodeBlockProgressiveDC(
+            PdfJsFrameComponent component,
+            ref short blockDataRef,
+            ref PdfJsHuffmanTable dcTable)
+        {
+            if (this.spectralEnd != 0)
+            {
+                throw new ImageFormatException("Can't merge DC and AC.");
+            }
+
+            this.CheckBits();
+
+            if (this.successiveHigh == 0)
+            {
+                // First scan for DC coefficient, must be first
+                int t = this.DecodeHuffman(ref dcTable);
+                int diff = t != 0 ? this.ExtendReceive(t) : 0;
+
+                int dc = component.DcPredictor + diff;
+                component.DcPredictor = dc;
+
+                blockDataRef = (short)(dc << this.successiveLow);
+            }
+            else
+            {
+                // Refinement scan for DC coefficient
+                if (this.GetBit() != 0)
+                {
+                    blockDataRef += (short)(1 << this.successiveLow);
+                }
+            }
+        }
+
+        private void DecodeBlockProgressiveAC(
```

Comment:
The method is quite long, might be worth to refactor internal branches into separate methods.

Reply:
I'll split out the Refinement part into a separate method.